### PR TITLE
[SYCL] Add macro to disable range rounding for FPGA compilations

### PR DESCRIPTION
**Pull request by smanna12**
Set SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING macro for           
1. both host and device compilations if -fsycl-disable-range-rounding flag is used.
2. all targets and HOST as well if the target (or one of targets) is FPGA.

        
This patch fixes the problems caused by https://github.com/intel/llvm/pull/3428 where it disables range rounding for FPGA target on device, 
but doesn't do the same for host compilation, which results in different assumptions about device code
between RT and compiler.

co-authored-by: Michael D Toguchi <michael.d.toguchi@intel.com>

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (mdtoguchi) [inline clang/lib/Frontend/InitPreprocessor.cpp:490]: This statement of `all FPGA compilations` is not true if you are only checking against `-fintelfpga`.  Using the triple is valid for FPGA as well (using `-fiintelfpga` will also imply the spir64_fpga triple)

Comment (smanna12) [reply clang/lib/Frontend/InitPreprocessor.cpp]: Thanks @mdtoguchi.  I have updated comments since I am  setting __SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING__ macro for  both host and device compilations if -fintelfpga flag is present.

File: clang/test/Preprocessor/predefined-macros.c
Comment (premanandrao) [inline clang/test/Preprocessor/predefined-macros.c:219]: Could you keep the spir64_fpga triple but without -fintelfpga and CHECK-NO-RANGE test?   This will confirm that we are changing the behavior from before.  i.e. keep the old test, but do the CHECK-NO-RANGE check.

The new variant with spir64 and -fintelfpga can be a separate RUN line in addition.

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (elizabethandrews) [inline clang/lib/Frontend/InitPreprocessor.cpp:1172]: Is -fintel-fpga always passed for FPGA compilations?

File: clang/test/Preprocessor/predefined-macros.c
Comment (smanna12) [reply clang/test/Preprocessor/predefined-macros.c:219]: Done. Thanks @premanandrao

**Comment by romanovvlad** [comment]
Is it possible that an application is compiled with `clang++ -fsycl -fintelfpga -fsycl-targets=spir64...;fpga` ? If so, could you please clarify, if the macro is defined for device compilation to SPIRV path?

**Comment by AGindinson** [comment]
> Is it possible that an application is compiled with `clang++ -fsycl -fintelfpga -fsycl-targets=spir64...;fpga` ?

The `-fsycl-targets` argument is obsolete in such a command line, as '-fintelfpga' effects include all `spir64_fpga` requirements. That said, I presume that `-fsycl-targets=spir64_fpga*` _only_ should still have the range rounding disabled under this macro.

**Review by premanandrao — COMMENTED** [review]
@smanna12, I am fine with the changes.   Could you add a test addressing @AGindinson's comment?

**Comment by smanna12** [comment]
> > Is it possible that an application is compiled with `clang++ -fsycl -fintelfpga -fsycl-targets=spir64...;fpga` ?
> 
> The `-fsycl-targets` argument is obsolete in such a command line, as '-fintelfpga' effects include all `spir64_fpga` requirements. That said, I presume that `-fsycl-targets=spir64_fpga*` _only_ should still have the range rounding disabled under this macro.

Thanks @romanovvlad and @AGindinson for looking the PR.

I tried the test case below. It does not set the macro with `-fsycl-targets=spir64_fpga*`. Should i use different test case?
```
+// RUN: %clang %s -E -dM -fsycl \
+// RUN: -fsycl-targets=spir64_fpga-unknown-unknown-sycldevice -o - \
+// RUN: | FileCheck -match-full-lines %s --check-prefix=CHECK-RANGE
+// CHECK-RANGE: #define __SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING__ 1
```

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (AaronBallman) [inline clang/lib/Frontend/InitPreprocessor.cpp:493]:
```suggestion
    if (LangOpts.SYCLAllowFpgaFlag)
      Builder.defineMacro("__SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING__");
```
The `1` is a default value anyway.

File: clang/include/clang/Basic/LangOptions.def
Comment (smanna12) [reply clang/include/clang/Basic/LangOptions.def]: Done

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (smanna12) [reply clang/lib/Frontend/InitPreprocessor.cpp]: Done

**Comment by smanna12** [comment]
> > Is it possible that an application is compiled with `clang++ -fsycl -fintelfpga -fsycl-targets=spir64...;fpga` ?
> 
> The `-fsycl-targets` argument is obsolete in such a command line, as '-fintelfpga' effects include all `spir64_fpga` requirements. That said, I presume that `-fsycl-targets=spir64_fpga*` _only_ should still have the range rounding disabled under this macro.

Thanks @AGindinson and @romanovvlad for taking a look at the PR.

@AGindinson 

> > > Is it possible that an application is compiled with `clang++ -fsycl -fintelfpga -fsycl-targets=spir64...;fpga` ?
> > 
> > 
> > The `-fsycl-targets` argument is obsolete in such a command line, as '-fintelfpga' effects include all `spir64_fpga` requirements. That said, I presume that `-fsycl-targets=spir64_fpga*` _only_ should still have the range rounding disabled under this macro.
> 
> Thanks @romanovvlad and @AGindinson for looking the PR.
> 
> I tried the test case below. It does not set the macro with `-fsycl-targets=spir64_fpga*`. Should i test different test cases here?
> 
> ```
> +// RUN: %clang %s -E -dM -fsycl \
> +// RUN: -fsycl-targets=spir64_fpga-unknown-unknown-sycldevice -o - \
> +// RUN: | FileCheck -match-full-lines %s --check-prefix=CHECK-RANGE
> +// CHECK-RANGE: #define __SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING__ 1
> ```

FYI, this part does not work yet.
We need to disable range rounding for host and target for all compilations if FPGA is involved (even with multiple targets).  @mdtoguchi is working on driver patch to provide additional information passed by the driver to control the setting of the macro with multiple targets .

**Review by premanandrao — APPROVED** [review]
LGTM!

@AaronBallman ?

File: clang/include/clang/Basic/LangOptions.def
Comment (AaronBallman) [inline clang/include/clang/Basic/LangOptions.def:255]:
```suggestion
LANGOPT(AheadOfTimeFPGACompilation, 1, 0, "Enables ahead-of-time compilation for FPGA")
```

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (AaronBallman) [inline clang/lib/Frontend/InitPreprocessor.cpp:493]:
```suggestion
    if (LangOpts.AheadOfTimeFPGACompilation)
      Builder.defineMacro("__SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING__");
```
Drops the braces as part of the usual community coding guidelines.

**Review by elizabethandrews — COMMENTED** [review]
I'm a little confused by this change. Please correct me if I am mistaken but it is my understanding that we need this macro defined for all FPGA compilations. 

While this change fixes the issue of inconsistency between host and device compilation because we now set it based on a -fintel-fpga flag instead of target, this also means the macro will not be defined for FPGA compilations which do not pass -fintel-fpga. Can you confirm this is the required behavior?

**Comment by smanna12** [comment]
> I'm a little confused by this change. Please correct me if I am mistaken but it is my understanding that we need this macro defined for all FPGA compilations.
> 
> While this change fixes the issue of inconsistency between host and device compilation because we now set it based on a -fintel-fpga flag instead of target, this also means the macro will not be defined for FPGA compilations which do not pass -fintel-fpga. Can you confirm this is the required behavior?

Yes, this is the required behavior. We will set the macro only with -fintelfpga flag on host and device compilation.

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (mdtoguchi) [inline clang/lib/Frontend/InitPreprocessor.cpp:491]: Shouldn't this now correlate with the `-fsycl-disable-range-rounding` setting given the new driver changes?

Comment (mdtoguchi) [reply clang/lib/Frontend/InitPreprocessor.cpp:491]: The scope that `-fintelfpga` covers is not enough, as it misses the cases when `-fsycl-targets=spir64_fpga` is used.  We should be able to use just `-fsycl-disable-range-rounding` to set the macro as the driver is taking care of determining all of the instances in which it should be set (whenever FPGA is used)

**Comment by premanandrao** [comment]
> Yes, this is the required behavior. We will set the macro only with -fintelfpga flag on host and device compilation.

I don't think this is true any more.  Please see @mdtoguchi's comments below.   We should be able to use just the -fsycl-disable-range-rounding flag to set the macro.

**Comment by smanna12** [comment]
> > Yes, this is the required behavior. We will set the macro only with -fintelfpga flag on host and device compilation.
> 
> I don't think this is true any more. Please see @mdtoguchi's comments below. We should be able to use just the -fsycl-disable-range-rounding flag to set the macro.

Yes,  this is not true anymore based on new driver's change.

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (smanna12) [reply clang/lib/Frontend/InitPreprocessor.cpp:491]: Thanks @mdtoguchi for the clarification. I will update patch.

**Comment by keryell** [comment]
@lforg37 we need to adapt this.

**Comment by smanna12** [comment]
buildbot/Lit_With_Cuda build is failing with error : 
FAILED: tools/libclc/CMakeFiles/builtins.link.libspirv-nvptx64--nvidiacl.dir/ptx-nvidiacl/libspirv/images/image.bc 
/localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./bin/clang -D_DEBUG -D_GNU_SOURCE -D__CLC_INTERNAL -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/libclc -I/localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.src/libclc -Iinclude -I/localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.src/llvm/include -I/localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.src/libclc/generic/include -target nvptx64--nvidiacl -Xclang -target-feature -Xclang +ptx64 -fno-builtin -nostdlib -I /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.src/libclc/ptx-nvidiacl/libspirv/images -o tools/libclc/CMakeFiles/builtins.link.libspirv-nvptx64--nvidiacl.dir/ptx-nvidiacl/libspirv/images/image.bc -c /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.src/libclc/ptx-nvidiacl/libspirv/images/image.cl -emit-llvm
/localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.src/libclc/ptx-nvidiacl/libspirv/images/image.cl:192:9: error: unknown type name 'half4'
typedef half4 pixelf16;
        ^
1 error generated.

This seems unrelated to the patch here.

http://ci.llvm.intel.com:8010/#/builders/37/builds/8857

**Comment by smanna12** [comment]
> > > Is it possible that an application is compiled with `clang++ -fsycl -fintelfpga -fsycl-targets=spir64...;fpga` ?
> > 
> > 
> > The `-fsycl-targets` argument is obsolete in such a command line, as '-fintelfpga' effects include all `spir64_fpga` requirements. That said, I presume that `-fsycl-targets=spir64_fpga*` _only_ should still have the range rounding disabled under this macro.
> 
> Thanks @AGindinson and @romanovvlad for taking a look at the PR.
> 
> @AGindinson
> 
> > > > Is it possible that an application is compiled with `clang++ -fsycl -fintelfpga -fsycl-targets=spir64...;fpga` ?
> > > 
> > > 
> > > The `-fsycl-targets` argument is obsolete in such a command line, as '-fintelfpga' effects include all `spir64_fpga` requirements. That said, I presume that `-fsycl-targets=spir64_fpga*` _only_ should still have the range rounding disabled under this macro.
> > 
> > 
> > Thanks @romanovvlad and @AGindinson for looking the PR.
> > I tried the test case below. It does not set the macro with `-fsycl-targets=spir64_fpga*`. Should i test different test cases here?
> > ```
> > +// RUN: %clang %s -E -dM -fsycl \
> > +// RUN: -fsycl-targets=spir64_fpga-unknown-unknown-sycldevice -o - \
> > +// RUN: | FileCheck -match-full-lines %s --check-prefix=CHECK-RANGE
> > +// CHECK-RANGE: #define __SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING__ 1
> > ```
> 
> FYI, this part does not work yet.
> We need to disable range rounding for host and target for all compilations if FPGA is involved (even with multiple targets). @mdtoguchi is working on driver patch to provide additional information passed by the driver to control the setting of the macro with multiple targets .

Done.

**Comment by smanna12** [comment]
Can I get a code-review please @AGindinson and FE code owners: @premanandrao , @AaronBallman or @elizabethandrews ? Thanks.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp:4615]: As per the [naming convention](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly):
```suggestion
    bool HasFPGA = false;
```

**Review by AGindinson — COMMENTED** [review]
The Driver part LGTM apart from a minor comment:

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (smanna12) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Thanks @AGindinson for caching this. Done

**Review by premanandrao — DISMISSED** [review]
FE changes look good to me.

**Review by elizabethandrews — DISMISSED** [review]
LGTM. Thanks!

**Review by AGindinson — APPROVED** [review]

**Comment by smanna12** [comment]
Thank you everyone for the reviews.

File: clang/include/clang/Basic/LangOptions.def
Comment (AaronBallman) [inline clang/include/clang/Basic/LangOptions.def:266]:
```suggestion
LANGOPT(SYCLDisableRangeRounding, 1, 0, "Disable parallel for range rounding")
```

File: clang/include/clang/Driver/Options.td
Comment (AaronBallman) [inline clang/include/clang/Driver/Options.td:5772]:
```suggestion
  HelpText<"Disable parallel for range rounding.">,
```

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (AaronBallman) [inline clang/lib/Driver/ToolChains/Clang.cpp:4622]: I am pretty sure this can be replaced by a call to `llvm::any_of()` instead of doing the loop manually (I'm not quite certain of the lambda signature for the predicate though as I'm not familiar with whatever the offload toolchain iterator is vending).

**Review by AaronBallman — COMMENTED** [review]
Just very minor nits from me.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (smanna12) [reply clang/lib/Driver/ToolChains/Clang.cpp:4622]: I am familiar to this Driver area. This was added by @mdtoguchi. I have created https://github.com/intel/llvm/issues/3784 for this. @AaronBallman, can we do this in a separate PR?

Comment (AaronBallman) [reply clang/lib/Driver/ToolChains/Clang.cpp:4622]: I'm fine handling it in a separate PR, it shouldn't be overly hard (it's an NFC change).

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (smanna12) [reply clang/lib/Driver/ToolChains/Clang.cpp:4622]: Thanks @AaronBallman. I have assigned the issue to myself and i will take a look at this after this PR lands.

**Comment by smanna12** [comment]
> LGTM!

Thank you.

**Review by mdtoguchi — APPROVED** [review]

**Comment by smanna12** [comment]
Thanks Mike.

**Comment by smanna12** [comment]
@bader, could you please merge this PR when you can? Thanks.

**Comment by smanna12** [comment]
Thank you @v-klochkov.

**Pull request merged**
